client/client_test.go

```diff
@@ -0,0 +1,190 @@
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/rs/zerolog"
+)
+
+func TestNew(t *testing.T) {
+	logger := zerolog.New(os.Stdout)
+	ctx := context.Background()
+
+	tests := []struct {
+		name        string
+		spec        *Spec
+		wantErr     bool
+		expectedErr error
+	}{
+		{
+			name:        "missing org",
+			spec:        &Spec{},
+			wantErr:     true,
+			expectedErr: ErrOrgRequired,
+		},
+		{
+			name: "missing app_id",
+			spec: &Spec{
+				Org: "test-org",
+			},
+			wantErr:     true,
+			expectedErr: ErrAppIDRequired,
+		},
+		{
+			name: "invalid app_id",
+			spec: &Spec{
+				Org:   "test-org",
+				AppID: "not-a-number",
+			},
+			wantErr:     true,
+			expectedErr: &ParseError{},
+		},
+		{
+			name: "missing installation_id",
+			spec: &Spec{
+				Org:   "test-org",
+				AppID: "12345",
+			},
+			wantErr:     true,
+			expectedErr: ErrInstallationIDRequired,
+		},
+		{
+			name: "invalid installation_id",
+			spec: &Spec{
+				Org:            "test-org",
+				AppID:          "12345",
+				InstallationID: "not-a-number",
+			},
+			wantErr:     true,
+			expectedErr: &ParseError{},
+		},
+		{
+			name: "missing private key",
+			spec: &Spec{
+				Org:            "test-org",
+				AppID:          "12345",
+				InstallationID: "67890",
+			},
+			wantErr:     true,
+			expectedErr: ErrPrivateKeyRequired,
+		},
+		{
+			name: "invalid private key format",
+			spec: &Spec{
+				Org:            "test-org",
+				AppID:          "12345",
+				InstallationID: "67890",
+				PrivateKey:     "invalid-key-content",
+			},
+			wantErr:     true,
+			expectedErr: ErrInvalidPrivateKeyFormat,
+		},
+		{
+			name: "valid config with private_key",
+			spec: &Spec{
+				Org:            "test-org",
+				AppID:          "12345",
+				InstallationID: "67890",
+				PrivateKey:     "-----BEGIN RSA PRIVATE KEY-----\ntest-content\n-----END RSA PRIVATE KEY-----",
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client, err := New(ctx, logger, tt.spec)
+
+			if tt.wantErr {
+				if err == nil {
+					t.Errorf("New() expected error but got none")
+					return
+				}
+
+				// Use typed error checking instead of string matching
+				if tt.expectedErr != nil {
+					switch expectedErr := tt.expectedErr.(type) {
+					case *ParseError:
+						var parseErr *ParseError
+						if !errors.As(err, &parseErr) {
+							t.Errorf("New() error = %v, expected ParseError type", err)
+						}
+					case error:
+						if !errors.Is(err, expectedErr) {
+							t.Errorf("New() error = %v, expected %v", err, expectedErr)
+						}
+					}
+				}
+			} else {
+				if err != nil {
+					t.Errorf("New() unexpected error = %v", err)
+				}
+				if client.ID() != "github-languages" {
+					t.Errorf("Client.ID() = %v, want %v", client.ID(), "github-languages")
+				}
+				if client.Org() != tt.spec.Org {
+					t.Errorf("Client.Org() = %v, want %v", client.Org(), tt.spec.Org)
+				}
```

Comment:
In what scenarios would we throw these errors?

Reply:
So if someone accidentally changed the client id here: https://github.com/guardian/cq-source-github-languages/blob/0c0177d783310ef1e27494f8b4f4debc98596fc8/client/client.go#L22 would be an example of when the first test would fail. For the second, if the spec wasn't retrieved correctly this would fail, say if this line had been changed: https://github.com/guardian/cq-source-github-languages/blob/0c0177d783310ef1e27494f8b4f4debc98596fc8/client/client.go#L30